docs: local harbor registry (#1877)
spectrocloud/librarium: docs/docs-content/integrations/harbor-edge.md

Author: This pack is called Harbor Edge-Native Config though. I think I am only using the word native when i am referring to  the pack itself

| `harbor-config.service.harbor` | Specifies the ports that harbor is exposed on. |
| `harbor-config.storage` | Specifies the size of the Harbor's persistent volume claim in GB. You can configure persistent volume claims for `registry`, `jobService`, `database`, `redis`, and `trivy`. |
| `harbor-config.service.metrics.enabled` | Specify whether to enable metrics on Harbor. For more information about the kinds of metrics that are exposed, refer to [Harbor metrics documentation](https://goharbor.io/docs/main/administration/metrics/). |

### Usage

#### Enable Harbor to Protect Against Outage
You can use Harbor in an Edge cluster that is connected to external networks. Harbor stores all container images downloaded from the internet and future image pulls from the cluster will be from the local harbor registry. If your cluster experiences an internet outage, it can still reboot containers or add new nodes using images stored locally in Harbor. For more information, refer to [Deploy a Cluster with a Local Harbor Registry](../clusters/edge/networking/local-registry.md).

#### Log in to Harbor Web UI

> With Harbor enabled on your Edge cluster, you can log in to Harbor via its web UI. By default, the Harbor service accessible as a NodePort-type service. with the HTTPS port being 30003. You can use the following steps to log in to Harbor via the web UI. 

Reviewer: ```suggestion
With Harbor enabled on your Edge cluster, you can log in to Harbor via its web UI. By default, the Harbor service is accessible as a NodePort-type service with HTTPS enabled on port `30003`. You can use the following steps to log in to Harbor via the web UI. 
```

Minor touches to the paragraph. It was a bit awkward to read.

1. Log in to [Palette](https://console.spectrocloud.com).

2. From the left **Main Menu**, navigate to **Clusters**. Choose your Edge cluster.

3. Navigate to the **Nodes** tab, in the **Private Ips** column, you can find the IP addresses of your Edge hosts. 

> 4. Open a new tab in your browser and navigate to `https://NODE_IP:30003` and replace `NODE_IP` with any IP address in your cluster. NodePort-type services are exposed on the same port on all nodes in your cluster. If you changed the HTTPS port in the configurations, replace the port with the HTTPS port you used. 

Reviewer: Lenny, should we call out that the user mush have network access to the network CIDR exposed by the Edge host? I'm just thinking of scenarios where an Edge host is in a separate network from the user, which is probably pretty common.

Author: i also thought about calling this out explicitly, but didn't because i felt this would be pretty evident. after some reflection so i will still add it

> 5. If you didn't provide a certificate or used a self-signed certificate, your browser might warn you about an unsafe connection. Dismiss the warning and you will be directed to Harbor's web UI. If you are using chrome, you can click anywhere in your browser tab and type "thisisunsafe" using your keyboard to dismiss the warning. 

Reviewer: ```suggestion
5. If you didn't provide a certificate or used a self-signed certificate, your browser might warn you about an unsafe connection. Dismiss the warning, and you will be directed to Harbor's web UI. If you use Google Chrome, you can click anywhere in your browser tab and type `this is unsafe` using your keyboard to dismiss the warning. 
```

I think putting `thisisunsafe` into code ticks makes it easier to read. Oh, and it's a single word.

#### Retrieve Harbor Credentials

During cluster creation, Palette creates two secrets that store the password used to authenticate the `admin` user in Harbor and the X509 certificate used for TLS. You can access both of these credentials using the following steps:

1. Configure kubectl to access your Kubernetes cluster. For more information, refer to [Access a Cluster with CLI](../clusters/cluster-management/palette-webctl.md). 

2. Depending on the credential you want to retrieve, use the commands described below. 

<Tabs>
<TabItem value="password" label="Password">

1. Issue the command `kubectl get namespaces` to get all namespaces. Look for a namespace called `cluster-CLUSTER-ID`, where `CLUSTER-ID` is a string of alphanumeric characters. This is the namespace where the secret for your Harbor password is stored. 

2. Issue the following command to get the password of your Harbor user. Replace `CLUSTER-ID-NAMESPACE` with the namespace you identified in the previous step. This command outputs your password. 

```shell

> kubectl get secret registry-info --namespace CLUSTER-ID-NAMESPACE -o jsonpath="{.data.SPECTRO_USER_PASSWORD}" | base64 --decode

Reviewer: ```suggestion
kubectl get secret registry-info --namespace CLUSTER-ID-NAMESPACE --output jsonpath="{.data.SPECTRO_USER_PASSWORD}" | base64 --decode
```
Adding long-form of `-o`

```

</TabItem>

<TabItem value="certificate" label="X509 certificate">
Issue the following command to retrieve the certificate. 

```shell

> kubectl get secret harbor-tls --namespace harbor -o jsonpath="{.data.tls\.crt}" | base64 --decode

Reviewer: ```suggestion
kubectl get secret harbor-tls --namespace harbor --output jsonpath="{.data.tls\.crt}" | base64 --decode
```

```

</TabItem>

<TabItem value="private-key" label="Private key">
Issue the following command to retrieve the private key.

```shell

> kubectl get secret harbor-tls --namespace harbor -o jsonpath="{.data.tls\.key}" | base64 --decode

Reviewer: ```suggestion
kubectl get secret harbor-tls --namespace harbor --output jsonpath="{.data.tls\.key}" | base64 --decode
```

:::info
If you didn't provide a certificate or are using a self-signed certificate, Docker will refuse to connect to the registry unless you configure Docker to trust the certificate authority or use a insecure connection for your Harbor registry. You can configure Docker to use a insecure connection by adding a line `"insecure-registries": ["REGISTRY_URL"]` in your Docker `daemon.json` file. For more information about `daemon.json`, refer to [Docker documentation](https://docs.docker.com/config/daemon/). 
:::

1. Log in to [Palette](https://console.spectrocloud.com).

2. From the left **Main Menu**, navigate to **Clusters**. Choose your Edge cluster.

3. Navigate to the **Nodes** tab, in the **Private Ips** column, you can find the IP addresses of your Edge hosts. 

4. Log in to Harbor from your command-line interface. The following example uses Docker, but you can use any other image management tool. Replace `NODE_IP` with the IP address of any of the nodes and replace `HARBOR_PASSWORD` with the password of your Harbor user. If you don't know your password, refer to [Retrieve Harbor Credentials](#retrieve-harbor-credentials). 

   ```shell

>    docker login NODE_IP:30003 -u admin -p HARBOR_PASSWORD

Reviewer: ```suggestion
   docker login NODE_IP:30003 --user admin --password HARBOR_PASSWORD
```

---
sidebar_label: 'harbor-edge'
title: 'Harbor Edge Native Config'
description: 'Harbor Edge Native Config pack in Spectro Cloud'
hide_table_of_contents: true
type: "integration"

> category: ['system-app', 'amd64']

Author: Arm64 has not been tested and rishi told me not to claim support for it. However, i think it probably does not need to be tagged as amd64 as we might support arm in the future.
